CRM/Civicaseextras/CaseDurationLog.php

```diff
@@ -0,0 +1,388 @@
+   *
+   * @param CRM_Core_DAO $caseActivity
+   *   Object with the data for the case and activity.
+   *
+   * @return bool
+   *   True if the current activity is closing the case, false otherwise
+   */
+  private function wasCaseClosing($caseActivity) {
+    $activityType = CRM_Utils_Array::value($caseActivity->activity_type_id, $this->activityTypes);
+
+    if ($activityType === 'Change Case Status') {
+      if ($this->extractStatusFromSubject($caseActivity->subject) === 'Closed') {
+        return TRUE;
+      }
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * Given an activity's subject, it tries to extract from the new case status,
+   * if it was opened or closed.
+   *
+   * @param string $activitySubject
+   *
+   * @return string|null
+   *   Returns 'Opened' or 'Closed' if it is able to extract the case status from
+   *   the subject. Otherwise it returns null.
+   */
+  private function extractStatusFromSubject($activitySubject) {
+    if (stripos($activitySubject, 'Case status changed from') !== FALSE) {
+      $subjectData = explode(' ', $activitySubject);
+      $newCaseStatus = $subjectData[sizeof($subjectData) - 1];
+
+      if (isset($this->statusGroupingsPerLabel[$newCaseStatus])) {
+        return $this->statusGroupingsPerLabel[$newCaseStatus];
+      }
+    }
+
+    return NULL;
+  }
+
+  /**
+   * Calculates the case's duration and stores it in the duration custom field.
+   *
+   * @param $caseID
+   */
+  public function calculateCaseDuration($caseID) {
+    $statsGroup = $this->getStatsGroup();
+    $durationField = $this->getDurationField($statsGroup);
+
+    $query = "
+      SELECT case_id, SUM(CASE WHEN duration IS NULL THEN datediff(CURDATE(), start_date) ELSE duration END) AS duration
+      FROM civicrm_case_duration_log
+      WHERE case_id = %1
+      GROUP BY case_id
+    ";
+    $params = array(1 => array($caseID, 'Integer'));
+    $durationResult = CRM_Core_DAO::executeQuery($query, $params);
+    $durationResult->fetch();
+
+    civicrm_api3('Case', 'create', array(
+      'custom_' . $durationField['id'] => $durationResult->duration,
+      'id' => 111,
+    ));
+  }
+
+  /**
+   * Calculates duration for all cases and stores it in the corresponding field.
+   */
+  public function calculateAllCasesDuration() {
+    $statsGroup = $this->getStatsGroup();
+    $durationField = $this->getDurationField($statsGroup);
+
+    CRM_Core_DAO::executeQuery('TRUNCATE TABLE ' . $statsGroup['table_name']);
+    CRM_Core_DAO::executeQuery("
+      INSERT INTO {$statsGroup['table_name']} (entity_id, {$durationField['column_name']})
+        SELECT case_id, SUM(CASE WHEN duration IS NULL THEN datediff(CURDATE(), start_date) ELSE duration END) AS duration
+        FROM civicrm_case_duration_log
+        GROUP BY case_id
+    ");
+  }
+
+  /**
+   * Obtains data for the custom group where case stats are stored.
+   *
+   * @return array
+   */
+  private function getStatsGroup() {

```

Comment:
As we talked about before, a dedicated class for fetching custom groups and fields might be useful here